ops/storage.py

```diff
@@ -62,6 +63,8 @@ def __init__(self, filename: Union['Path', str]):
         self._db = sqlite3.connect(str(filename),
                                    isolation_level=None,
                                    timeout=self.DB_LOCK_TIMEOUT.total_seconds())
+        if filename != ":memory:":
```

Comment:
Offhand, is there any reason why a failure of this command (for any reason), should cause it to fail?
I wonder if we want to wrap it with at try/finally.

Officially, if we *really* care about the security of this, this particular framing creates a short window of security hole, where the file is created, someone could see it, open it, and then it gets chmod away. (This was the feedback from the Security team for our juju-run socket.)

If we want it to be correct, the correct action is to do something like:
```
    os.close(os.open(path, flags, 0o600))
    self._db = sqlite3.connect(...)
```
So we always create the file with the right perms.
I don't know that it is worth that overhead, but it does eliminate any form of race condition if we genuinely care about security.
(And we should also have a way for pre-existing files to make sure they get set correctly, so it is probably something like:
def _ensure_db_permissions(self, filename):
    filepth = pathlib.Path(filename)
    if filepth.exists():
      os.chmod()
    else:
      try:
        fd = os.open(..., 0o666)
      except ???
         ...
      try:
        ...
      finally:
         os.close(fd)
```

I don't think it is a particularly complicated function, but ensuring that you open files with secure permissions is pretty well know if we really want to be secure here.

Reply:
Yeah, I think that's reasonable, John -- thanks for the review. Might as well fix it properly and avoid the race condition. And let's put a try/except around the `chmod` as you suggest, and log the error instead of fail hard.

I believe the flags for the `os.open` call should be `os.open(filename, os.O_CREAT, mode=0o600)`.

Reply:
> Might as well fix it properly and avoid the race condition.

Fine with me. When we talked about this last week @benhoyt you thought that there was also a need to put the file in a secured directory (for handling temporary files that sqlite uses for atomicity, I think). I did some basic searching and couldn't find anything about this. Do you remember anything more specific, or should I ignore that for now?

> And let's put a try/except around the `chmod` as you suggest, and log the error instead of fail hard.

I'm not convinced about this. I could definitely be convinced that we should add a try/except and raise a clearer error. If we just log the error and don't hard fail, doesn't this effectively undermine the security? We're then relying on the admin to be checking the log and noticing this error if something goes wrong with securing the permissions. That doesn't feel right to me.

> I believe the flags for the `os.open` call should be `os.open(filename, os.O_CREAT, mode=0o600)`.

I think it needs to be `O_CREAT | O_EXCL`, because otherwise there's a small race window between the check if the file already exists and creating the new file. With `O_EXCL` if someone creates the file in that window then the open will fail.

Reply:
> I'm not convinced about this. I could definitely be convinced that we should add a try/except and raise a clearer error. If we just log the error and don't hard fail, doesn't this effectively undermine the security? We're then relying on the admin to be checking the log and noticing this error if something goes wrong with securing the permissions. That doesn't feel right to me.

Thinking a bit more: I could maybe be convinced that the already-exists case should not hard fail, since you're just not getting the "upgrade" in security if it fails. However, I am definitely not convinced that the new-file case should just log the error.

Reply:
The directory thing I mentioned was that the directory the sqlite file needs to be writeable, so sqlite can create the WAL file(s) in it. See https://sqlite.org/wal.html and search for "directory". We could try to pre-create those files too, but I think that's probably getting a bit carried away?

Regarding the exception handling: I think you're right: if we're going to layer on a bit of security, we should hard fail if it doesn't work (at least in the new-file case).

Good catch on the `O_EXCL` issue.